### ISSUE TITLE
Expose knex.cast function

### ISSUE THREAD

**Issue by wubzz** [knex/knex#2587]
# Feature discussion / request

1. Explain what is your use case
To be able to cast values and references to specific types.

2. Explain what kind of feature would support this
Exposing a new function (or rather a set of functions)

3. Give some API proposal, how the feature should work
Disclaimer: Mostly inspiried by objectionjs as per original discussion in https://github.com/tgriesser/knex/issues/2507

```javascript
knex.cast(value, type);
knex.castText(value);
knex.castInt(value);
knex.castBigInt(value);
knex.castFloat(value);
knex.castDecimal(value);
knex.castReal(value);
knex.castBool(value);
knex.castJson(value);
knex.cast*(value).as(alias)
```

Intended to support any "valid-knex" input:
```javsacript
knex.castText(1); //CAST '1' AS text
knex.castText(knex.raw(1)); //CAST 1 AS text
knex.castText(knex.ref('id')); // CAST "id" AS text
```

Example implementation
```javascript
'use strict';
import Raw from './raw';

class Cast extends Raw {
  constructor(client) {
    super(client);

    this.value = null;
    this.to    = '';
    this.alias = '';
  }

  cast(value, to) {
    this.value = value;
    this.to    = to;

    return this;
  }

  as(alias) {
    this.alias = alias;

    return this;
  }

  static get TYPES() {
    return {
      Text:    'text',
      Integer: 'integer',
      BigInt:  'bigint',
      Float:   'float',
      Decimal: 'decimal',
      Real:    'real',
      Bool:    'boolean',
      Json:    'json',
    };
  }

  castText(value) {
    return this.cast(value, Cast.TYPES.Text);
  }

  castInt(value) {
    return this.cast(value, Cast.TYPES.Integer);
  }

  castBigInt(value) {
    return this.cast(value, Cast.TYPES.BigInt);
  }

  castFloat(value) {
    return this.cast(value, Cast.TYPES.Float);
  }

  castDecimal(value) {
    return this.cast(value, Cast.TYPES.Decimal);
  }

  castReal(value) {
    return this.cast(value, Cast.TYPES.Real);
  }

  castBool(value) {
    return this.cast(value, Cast.TYPES.Bool);
  }

  castJson(value) {
    return this.cast(value, Cast.TYPES.Json);
  }

  toSQL() {
    const {value, to, alias} = this;

    const asAlias = alias ? `AS ${this.client.formatter(this).wrap(alias)}` : '';

    const cast = to === Cast.TYPES.Json
      ? `to_jsonb(?) ${asAlias}`
      : `CAST(? AS ${to}) ${asAlias}`;

    this.set(cast, [value]);

    return super.toSQL(...arguments);
  }
}

export default Cast
```

Extending `Raw` _(same as `Ref`)_ makes it a valid input across the knex API and adds automatic support for existing `Raw` functions such as `.wrap`.

Creating this for discussion purposes, I'm not sure what would be the most optimal API or implementation, but I believe this is something knex currently lacks.

**Comment by elhigu** [comment]
I think that casting should be allowed for column references too and not only for literal values. Have you tried if `knex.raw('?', [knex.ref('foo.bar')])` actually outputs column ref to replace the `?`. IIRC this actually should work for casting complete queries / raw too :+1:

**Comment by wubzz** [comment]
Yes the idea is that this covers all bases, including the new `.ref`. It plays well with existing query-compiler in that sense. For example `knex.castText(knex.ref('id')); // CAST "id" AS text`

What will not make sense and not even work is something like this, but it's kind of a given imo.. :)
```knex.castText(knex.ref('id').as('Test')).as('Test');```

**Comment by mercmobily** [comment]
Hi! Thanks for the thorough proposal and examples; agreed that a small, ergonomic cast helper would be very useful.

Quick summary of the current status:

This is in progress via `#5039` (Add cast functions to support 'cast as' statements). That PR introduces `knex.cast`‑style helpers, updates TypeScript typings, and has passing integration for PostgreSQL, SQLite3, MySQL, and CockroachDB. OracleDB and MSSQL still need work, and documentation is pending.

Current behaviour without the PR is to use `knex.raw` (or `knex.ref` with `raw`) for casts, for example `CAST(? AS integer)` with a binding of `knex.ref('id')`. There is no first‑class `knex.cast` helper in a released version yet, so adopting the PR will close this gap.

What `#5039` covers so far
- Adds cast helpers analogous to `knex.raw`/`knex.ref` so that `CAST ... AS ...` can be expressed portably.
- Confirms PostgreSQL integration; SQLite3, MySQL and CockroachDB are marked OK in the PR checklist.
- Updates typings in `types/index.d.ts` (the PR notes this is complete).

What remains before merge
- OracleDB: implement and verify dialect‑appropriate rendering for cast targets (e.g. string/number/blob variants) and add tests.
- MSSQL: investigate oddities when casting to `varbinary(max)` (the PR includes a screenshot of an unexpected result); adjust compiler or normalisation accordingly and add tests.
- Documentation: add a short section showing `knex.cast(value, type)` usage, including aliasing and a couple of per‑dialect notes (for example, when a `::type` shorthand is preferred in PostgreSQL).

Technical notes and considerations
- Dialect differences matter: PostgreSQL often prefers `::type` in addition to `CAST(... AS ...)`, MySQL/MariaDB use `CAST`, SQLite supports a subset, and Oracle has its own specifics. The PR routes through per‑dialect compilers so each can render idiomatically.
- Typings are in scope: the PR already touches `types/index.d.ts`, which is where return types and helper signatures should live. Keeping these crisp will prevent regressions similar to other aggregate/type helpers.
